### DISCUSSION TITLE
how to make rust-analyzer use the pipenv environment, and prefix its commands with env variable: LIBTORCH_USE_PYTORCH=1

### DISCUSSION THREAD

**Discussion by DeaDSouL** [rust-lang/rust-analyzer#16978] [Q&A] [answered]
Hello,

I use VSCode with Rust-Analyzer, for my work with rust. And I needed the libtorch package, but since my OS (Fedora) doesn't provide one with its repositories. I've created a virtual python environment in my rust project folder (via: `pipenv install torch==2.2.0`).
Now, in order to activate that environment I need to run `pipenv shell`.
Then I can run the following: `LIBTORCH_USE_PYTORCH=1 cargo run` and it works just fine.
I can also run a command inside that virtual environment without entering it via: `pipenv run the-command`.

so, I've configured my vscode terminal to use that environment, by configuring the `python.defaultInterpreterPath` in vscode.

now my problem is, how to tell rust-analyzer to use that environment, and prefix all its commands with this env variable `LIBTORCH_USE_PYTORCH=1` ?

I tried all its configurations, but no luck.

Thanks

**Comment by lnicola** [comment, marked as answer]
You can start Code in `pipenv shell` or set `rust-analyzer.server.extraEnv` in the extension settings ([example](https://rust-analyzer.github.io/manual.html#toolchain)).

**Reply by DeaDSouL** [reply]
I tried this:

```
{
..
  "rust-analyzer.runnables.command": "LIBTORCH_USE_PYTORCH=1 /home/deadsoul/.local/bin/pipenv run cargo",
  "rust-analyzer.server.extraEnv": {"LIBTORCH_USE_PYTORCH": "1"},
..
}
```

no luck!

**Reply by lnicola** [reply]
Yeah, for runnables you might have to run Code inside `pipenv`. I suspect having the environment variable set isn't enough, you also want `PATH`, `PYTHONPATH` or whatever.

**Reply by DeaDSouL** [reply]
Never mind, 
I tried running Code inside `pipenv` and it worked. I'll mark your first reply as the solution
Thanks

**Reply by lnicola** [reply]
Yeah, `extraEnv` can be useful if your code needs some environment variables for a build script, but nothing more fancy.
